File: src/main/java/player/ki/concurrent/MultiThreadStrategy.java
package player.ki.concurrent;

import board.ChessGameInterface;
import helper.Move;
import player.ki.AbstractComputerPlayerUI;
import player.ki.DynamicEvaluation;

import java.util.*;
import java.util.concurrent.*;

class MultiThreadStrategy extends AbstractConcurrencyStrategy {
    //  final private DelayedExecutorService delayedExecutorService;
    private final ExecutorService executorService;
    private final int numberOfThreads;

    MultiThreadStrategy(AbstractComputerPlayerUI ui, int numberOfCores) {
        super(ui);

        numberOfThreads = numberOfCores;
        executorService = Executors.newFixedThreadPool(numberOfCores);
//		delayedExecutorService = new DelayedExecutorService(numberOfCores);
    }


    @Override
    public SortedSet<EvaluatedMove> evaluatePossibleMoves(final ChessGameInterface game, final DynamicEvaluation dynamicEvaluation) {
        //as long as the first parameter is 0 and the second one is bigger
        //the progress bar will always show correctly 0% (so '1' as second parameter is fine)
        showProgress(0, 1);

        final LinkedList<Callable<EvaluatedMove>> movesToEvaluate = getEvaluatableMoves(game, dynamicEvaluation);

        SortedSet<EvaluatedMove> result = Collections.emptySortedSet();
        try {
            result = evaluate(movesToEvaluate);//resultFutures.get();
        } catch (Exception e) {
            e.printStackTrace();
        }

        assert !result.isEmpty() : "no evaluation of a possible moves was successfull";

        return result;
    }

    private SortedSet<EvaluatedMove> evaluate(final LinkedList<Callable<EvaluatedMove>> movesToEvaluate)
            throws InterruptedException, ExecutionException {
        SortedSet<EvaluatedMove> result = new TreeSet<EvaluatedMove>();

        final int totalNumberOfMoves = movesToEvaluate.size();

        CompletionService<EvaluatedMove> ecs = new ExecutorCompletionService<EvaluatedMove>(executorService);
        submitCallables(movesToEvaluate, ecs, numberOfThreads);

        for (int i = 0; i < totalNumberOfMoves; ++i) {
            //wait for an evaluation to be finished
            EvaluatedMove evaluatedMove = ecs.take().get();
            //show the progress
            showProgress(i, totalNumberOfMoves);
            //add a new move to be evaluated to the queue (if some are left)
            submitCallables(movesToEvaluate, ecs, 1);
            //add this evaluation to result set
            if (evaluatedMove != null) {
                result.add(evaluatedMove);
            }
        }

        return result;
    }

    private LinkedList<Callable<EvaluatedMove>> getEvaluatableMoves(final ChessGameInterface game, final DynamicEvaluation dynamicEvaluation) {
        final List<Move> possibleMoves = getPossibleMoves(game);
        assert !possibleMoves.isEmpty() : "no moves were possible and therefore evaluatable";

        final int totalNumberOfMoves = possibleMoves.size();

        final Iterator<ChessGameInterface> gameInstances = game.copyGame(totalNumberOfMoves).iterator();

        final LinkedList<Callable<EvaluatedMove>> movesToEvaluate = new LinkedList<Callable<EvaluatedMove>>();
        for (Move move : possibleMoves) {
            movesToEvaluate.add(
                    new MoveEvaluationCallable(
                            gameInstances.next(), move, dynamicEvaluation//, delayedExecutorService
                    )
            );
        }

        return movesToEvaluate;
    }

    private void submitCallables(LinkedList<Callable<EvaluatedMove>> movesToEvaluate, CompletionService<EvaluatedMove> completionService, int numberOfMovesToSubmit) {
        for (int i = 0; i < numberOfMovesToSubmit; i++) {
            if (movesToEvaluate.isEmpty()) {
                break;
            }
            completionService.submit(movesToEvaluate.removeFirst());
        }
    }

    private static class MoveEvaluationCallable implements Callable<EvaluatedMove> {
        private final DynamicEvaluation dynamicEvaluation;
        private final ChessGameInterface game;
        private final Move move;

        MoveEvaluationCallable(ChessGameInterface game, Move move, DynamicEvaluation dynamicEvaluation) {
            this.dynamicEvaluation = dynamicEvaluation;
            this.game = game;
            this.move = move;
        }

        public EvaluatedMove call() throws Exception {
            try {
                final float value = dynamicEvaluation.evaluateMove(game, move);
                EvaluatedMove evaluatedMove = new EvaluatedMove(move, value);

                return evaluatedMove;
            } catch (Exception e) {
                //print out the error
                e.printStackTrace();
                return null;
            }
        }
    }

}

File: src/main/kotlin/voidchess/player/ki/concurrent/MultiThreadStrategy.kt
package voidchess.player.ki.concurrent

import voidchess.board.ChessGameInterface
import voidchess.board.move.Move
import voidchess.player.ki.evaluation.EvaluatingMinMax
import voidchess.player.ki.evaluation.EvaluatedMove

import java.util.*
import java.util.concurrent.*
import kotlin.collections.ArrayList

internal class MultiThreadStrategy(showProgress: (Int, Int)->Unit, private val numberOfThreads: Int) : AbstractConcurrencyStrategy(showProgress) {
    private val executorService: ExecutorService = Executors.newFixedThreadPool(numberOfThreads)


    override fun evaluateMoves(game: ChessGameInterface, evaluatingMinMax: EvaluatingMinMax): MutableList<EvaluatedMove> {
        //as long as the first parameter is 0 and the second one is bigger
        //the progress bar will always show correctly 0% (so '1' as second parameter is fine)
        showProgress(0, 1)

        val movesToEvaluate = getEvaluableMoves(game, evaluatingMinMax)

        var result = try {
            evaluate(movesToEvaluate)
        } catch (e: Exception) {
            e.printStackTrace()
            LinkedList<EvaluatedMove>()
        }

        assert(result.isNotEmpty()) { "no evaluation of a possible moves was successful" }

        return result
    }

    @Throws(InterruptedException::class, ExecutionException::class)
    private fun evaluate(movesToEvaluate: LinkedList<Callable<EvaluatedMove>>): MutableList<EvaluatedMove> {
        val totalNumberOfMoves = movesToEvaluate.size
        val result = ArrayList<EvaluatedMove>(totalNumberOfMoves)

        val ecs = ExecutorCompletionService<EvaluatedMove>(executorService)
        submitCallables(movesToEvaluate, ecs, numberOfThreads)

        for (i in 0 until totalNumberOfMoves) {
            //wait for an evaluation to be finished
            val evaluatedMove = ecs.take().get()
            //show the progress
            showProgress(i, totalNumberOfMoves)
            //add a new move to be evaluated to the queue (if some are left)
            submitCallables(movesToEvaluate, ecs, 1)
            //add this evaluation to result set
            if (evaluatedMove != null) {
                result.add(evaluatedMove)
            }
        }

        return result
    }

    private fun getEvaluableMoves(game: ChessGameInterface, evaluatingMinMax: EvaluatingMinMax): LinkedList<Callable<EvaluatedMove>> {
        val possibleMoves = game.getAllMoves()
        assert(possibleMoves.isNotEmpty()) { "no moves were possible and therefore evaluable" }

        val totalNumberOfMoves = possibleMoves.size

        val gameInstances = game.copyGame(totalNumberOfMoves).iterator()

        val movesToEvaluate = LinkedList<Callable<EvaluatedMove>>()
        for (move in possibleMoves) {
            movesToEvaluate.add(
                    MoveEvaluationCallable(
                            gameInstances.next(), move, evaluatingMinMax
                    )
            )
        }

        return movesToEvaluate
    }

    private fun submitCallables(movesToEvaluate: LinkedList<Callable<EvaluatedMove>>, completionService: CompletionService<EvaluatedMove>, numberOfMovesToSubmit: Int) {
        for (i in 0 until numberOfMovesToSubmit) {
            if (movesToEvaluate.isEmpty()) {
                break
            }
            completionService.submit(movesToEvaluate.removeFirst())
        }
    }

    private class MoveEvaluationCallable internal constructor(private val game: ChessGameInterface, private val move: Move, private val evaluatingMinMax: EvaluatingMinMax) : Callable<EvaluatedMove> {

        @Throws(Exception::class)
        override fun call(): EvaluatedMove? {
            return try {
                val value = evaluatingMinMax.evaluateMove(game, move)
                EvaluatedMove(move, value)
            } catch (e: Exception) {
                //print out the error
                e.printStackTrace()
                null
            }
        }
    }

}
